apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,42 @@
+import asyncHandler from 'express-async-handler';
+import * as sqldb from '@prairielearn/postgres';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.user.user_id,
+    session_id: req.session.id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+    accept: req.headers['accept'],
+  };
+  // Need a better way to find or insert a fingerprint. Currently doing this through two calls but would probably be better to do it in one.
+  // First call, check if fingerprint exists in the database
+  let client_fingerprint_id: number | null =
+    parseInt(
+      (await sqldb.queryZeroOrOneRowAsync(sql.select_client_fingerprint, params)).rows[0]?.id,
+    ) ?? null;
```

Comment:
Because Javascript numbers are prone to loss of precision for very large IDs, we typically treat IDs as strings.

Reply:
Also, newer code uses `queryOptionalRow` instead of `queryZeroOrOneRowAsync`.

Reply:
I have updated this query to use `queryOptionalRow` in commit [b135adf](https://github.com/PrairieLearn/PrairieLearn/pull/8798/commits/b135adfd92d740104813013c22b9b36e61ec23bd). Also, I am now passing in the `ClientFingerprintSchema` and removing the use of `parseInt`.